R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+
+# ----------
+
+# # Current design decisions
+# * tagQuery objects or tag environments can NOT be used in UI. These objects MUST be converted back to standard tag objects.
+# * tagFunctions will not be altered in place
+#   * To alter tagFunction()s, use the `onRender(x)` method to register a method to be called after `as.tags(x)` is called.
+#   * `onRender(x, expr)` will wrap create a tag function that will resolve the tags before running the expr.
+
+
+## rlang::sexp_address()
+# Use to get a unique key for stacks
+# Use `env$envKey` over `rlang::sexp_address()`; 10x speed improvement
+
+# Use for `has()` functionality
+envirMap <- function() {
+  map <- fastmap()
+  list(
+    keys = map$keys,
+    asList = function() {
+      unname(map$as_list())
+    },
+    has = function(envir) {
+      map$has(envir$envKey)
+    },
+    set = function(envir, value = envir) {
+      map$set(envir$envKey, value)
+    },
+    remove = function(envir) {
+      map$remove(envir$envKey)
+    }
+  )
+}
+# Use for consistent `asList()` order
+envirStack <- function() {
+  stack <- faststack()
+  list(
+    push = stack$push,
+    asList = stack$as_list,
+    uniqueList = function() {
+      unique(stack$as_list())
+    }
+  )
+}
+
+# (Used for `unique_envirStack()` only. Do not use directly!)
+envirStackUni_ <- function() {
+  map <- fastmap()
+  stack <- faststack()
+  list(
+    push = function(env) {
+      key <- env$envKey
+      if (!map$has(key)) {
+        # mark the key as _seen_
+        map$set(key, TRUE)
+        # add the env
+        stack$push(env)
+      }
+    },
+    uniqueList = stack$as_list
+  )
+}
+# Use to retrieve unique environments (eg: `tq$parent()`)
+envirStackUnique <- function() {
+  stack <- envirStack()
+  count <- 0
+  list(
+    push = function(env) {
+      count <<- count + 1
+      if (count == 500) {
+        # convert the current stack to a `envirStackUni_()`
+        newStack <- envirStackUni_()
+        walk(stack$asList(), newStack$push)
+        stack <<- newStack
+      }
+      stack$push(env)
+    },
+    uniqueList = function() {stack$uniqueList()}
+  )
+}
+
+
+
+
+
+
+# Retrieve all attributes that can be manually set
+# ?attr
+# Note that some attributes (namely ‘class’, ‘comment’, ‘dim’,
+# ‘dimnames’, ‘names’, ‘row.names’ and ‘tsp’) are treated specially
+# and have restrictions on the values which can be set.
+safeAttrValues <- function(x) {
+  badElAttrs <- c("class", "comment", "dim", "dimnames", "names", "row.names", "tsp")
+  attrVals <- attributes(x)
+  attrVals[badElAttrs] <- NULL
+  attrVals
+}
+
+# Convert a list to an environment and keep class and attribute information
+safeListToEnv <- function(x, newClass = NULL) {
+  xList <- x
+
+  ret <- list2env(xList, new.env(parent = emptyenv()))
+
+  attrVals <- safeAttrValues(xList)
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- c(newClass, class(xList), "environment")
```

Comment:
Calling `class(xList)` on a plain list will return `"list"`, so the result may have the class `c("list", "environment")`.

I think that using `oldClass()` makes sense here, though I have to admit it's not 100% clear to me exactly what that function is supposed to do.

Reply:
From https://stat.ethz.ch/pipermail/r-help/2005-August/076881.html:

> so class returns the implicit class, and oldClass does not.

Reply:
Also, I'm curious: is the explicit `"environment"` class needed?

Reply:
`oldClass(x)`... TIL

Reply:
> is the explicit `"environment"` class needed?

Not necessarily. But Hadley liked to add the class `"function"` to things like `shiny::reactive()` where it isn't necessary?  I was following his pattern.

Reply:
After discussion, we decided to remove it.